File: java/easy_1/two_number_sum/TwoNumberSum.java
package main.java.com.example.twonumbersum;

import java.util.HashMap;

// O(n) time | O(n) space
public class TwoNumberSum {
    
    /**
     * Finds two numbers in the array that add up to the targetSum.
     *
     * @param array The array of integers to search within.
     * @param targetSum The target sum of the two numbers to find.
     * @return An array containing the two numbers that add up to targetSum, or an empty array if no such pair exists.
     */
    public static int[] twoNumberSum(int[] array, int targetSum) {
        // HashMap to store numbers we've seen so far
        HashMap<Integer, Boolean> map = new HashMap<>();

        // Iterate through each number in the array
        for (int num : array) {
            // Calculate the difference needed to reach targetSum
            int diff = targetSum - num;
            
            // Check if the difference is already in the HashMap
            if (map.containsKey(diff)) {
                // If found, return the pair of numbers that add up to targetSum
                return new int[] {num, diff};
            } else {
                // Otherwise, add the current number to the HashMap
                map.put(num, true);
            }
        }
        // Return an empty array if no pair is found
        return new int[] {};
    }
}

File: python/Easy/two_number_sum/two_number_sum.py
# O(n) time | O(n) space
def twoNumberSum(array, targetSum):
    """
    Finds two numbers in the array that add up to the targetSum.

    Parameters:
    array (list): List of integers.
    targetSum (int): The target sum to find in the array.

    Returns:
    list: A list containing the two numbers that add up to targetSum, or an empty list if no such numbers are found.
    """
    # Dictionary to store numbers we've seen so far
    dict = {}

    # Iterate through each number in the array
    for num in array:
        # Calculate the difference needed to reach targetSum
        diff = targetSum - num

        # Check if the difference is already in the dictionary
        if diff in dict:
            # If found, return the pair of numbers that add up to targetSum
            return [num, diff]
        else:
            # Otherwise, add the current number to the dictionary
            dict[num] = True

    # Return an empty list if no pair is found
    return []